File: src/main/go/src/euler/algorithm/prime.go
// COPYRIGHT (C) 2014 barreiro. All Rights Reserved.
// GoLang helper for Project Euler problems
package algorithm

import (
	"math"
	"math/bits"
)

var millerRabinFast, millerRabinBase = []int{2, 7, 61}, []int{2, 325, 9375, 28178, 450775, 9780504, 1795265022}

// calculates the prime factors of a given number. The result is a map where the keys are primes and the values are the occurrences
func PrimeFactors(n int) map[int]int {
	factorMap, generator, stop := make(map[int]int), GeneratorTrialDivision(), IntSqrt(n)

	for n != 1 {
		factor := generator()
		for ; n%factor == 0; n /= factor {
			factorMap[factor]++
		}

		if factor >= stop {
			// if the number is prime, or if there is still a remainder, add itself as a factor
			if n >= factor || len(factorMap) == 0 {
				factorMap[n] = 1
			}
			break
		}
	}
	return factorMap
}

func PrimeComposition(factormap map[int]int) int {
	sum := 1
	for base, exp := range factormap {
		sum *= Pow(base, exp)
	}
	return sum
}

// closure that generates primes based on the method of trial division
func GeneratorTrialDivision() func() int {

	var cache []int

	return func() int {
		// avoid put 2 into the cache. It's easy to skip even numbers altogether
		if cache == nil {
			cache = make([]int, 0)
			return 2
		}
		// 3 is the smallest value in the cache. It's put explicitly so that the last element can always be retrieved
		if len(cache) == 0 {
			cache = append(cache, 3)
			return 3
		}
		for candidate, stop := cache[len(cache)-1]+2, IntSqrt(cache[len(cache)-1]); ; candidate += 2 {
			small := candidate <= math.MaxInt32
			for _, prime := range cache {
				if small {
					if int32(candidate)%int32(prime) == 0 {
						break
					}
				} else if candidate%prime == 0 {
					break
				}

				if prime > stop {
					// won't find a prime factor, therefore the candidate is prime
					cache = append(cache, candidate)
					return candidate
				}
			}
		}
	}
}

// closure that generates primes prime numbers, starting with the one below N.
func PrimesLessThan(n int) func() int {

	candidate := n

	return func() int {
		for candidate--; ; candidate-- {
			if MillerRabin(candidate) {
				return candidate
			}
		}
	}
}

func MillerRabin(n int) bool {
	effectiveBase := millerRabinFast
	if n >= 4759123141 {
		effectiveBase = millerRabinBase
	}

	for _, b := range effectiveBase {
		if n > b && !millerRabinPass(b, n) {
			return false
		}
	}
	return true
	// return (n > 1) && ((n == 2) || stream(effectiveBase).allMatch(b- > n <= b || millerRabinPass(b, n)));
}

func millerRabinPass(b, n int) bool {
	s := bits.TrailingZeros64(uint64(n - 1))
	d := (n - 1) >> uint64(s)
	a := PowerMod(b, d, n)

	if a == 1 {
		return true
	}
	for i := 0; i < s-1; i++ {
		if a == n-1 {
			return true
		}
		if a == 1 {
			return false
		}
		a = PowerMod(a, 2, n)
	}
	return a == n-1
}

File: src/main/rust/euler/algorithm/prime.rs
// COPYRIGHT (C) 2017 barreiro. All Rights Reserved.
// Rust solvers for Project Euler problems

use std::collections::HashMap;

use euler::algorithm::long::{floor_sqrt, is_even};
use euler::algorithm::long::power_modulo;

// Bases for the Miller-Rabin test
const MR_THRESHOLD: isize = 4_759_123_141;
const MR_BASE_FAST: &[isize] = &[2, 7, 61];
const MR_BASE: &[isize] = &[2, 325, 9_375, 28_178, 450_775, 9_780_504, 1_795_265_022];

/// calculates the prime factors of a given number. The result is a map where the keys are primes and the values are the occurrences
pub fn prime_factors(n: isize) -> HashMap<isize, isize> {
    let (mut factor_map, mut value, small, stop) = (HashMap::new(), n, n <= i32::max_value() as _, floor_sqrt(n));
    for factor in generator_trial_division() {
        while if small { value as i32 % factor as i32 == 0 } else { value % factor == 0 } {
            value = if small { (value as i32 / factor as i32) as _ } else { value / factor };
            factor_map.entry(factor).and_modify(|e| *e += 1).or_insert(1);
        }
        if value == 1 {
            break;
        }
        if factor >= stop {
            // the number is prime or if there is still a remainder, add it as a factor
            factor_map.insert(value, 1);
            break;
        }
    }
    factor_map
}

// --- //

/// closure that generates primes based on the method of trial division
pub struct GeneratorTrialDivision {
    sieve: Vec<isize>
}

pub fn generator_trial_division() -> GeneratorTrialDivision {
    GeneratorTrialDivision { sieve: vec![] }
}

impl Iterator for GeneratorTrialDivision {
    type Item = isize;

    fn next(&mut self) -> Option<Self::Item> {
        let next = match self.sieve.last() {
            None => { 2 }
            Some(2) => {
                self.sieve.clear();
                3
            }
            last => { (last.unwrap() + 2..).step_by(2).find(|&candidate| prime_sieve(candidate, &self.sieve)).unwrap() }
        };

        self.sieve.push(next);
        Some(next)
    }
}

pub fn prime_sieve(n: isize, sieve: &[isize]) -> bool {
    let ceil = floor_sqrt(n);
    if n <= i32::max_value() as _ {
        !sieve.iter().take_while(|&&factor| factor <= ceil).any(|&f| n as i32 % f as i32 == 0)
    } else {
        !sieve.iter().take_while(|&&factor| factor <= ceil).any(|&f| n % f == 0)
    }
}

// --- //

/// closure that generates primes based on the method of wheel factorization
pub struct GeneratorWheel {
    size: isize,
    increments: Vec<isize>,
    sieve: Vec<isize>,
}

pub fn generator_wheel() -> GeneratorWheel {
    generator_custom_wheel(&[2, 3, 5, 7, 11])
}

pub fn generator_custom_wheel(primes: &[isize]) -> GeneratorWheel {
    let size = primes.iter().product();
    let buckets = (size..=1 + size * 2).filter(|n| primes.iter().all(|p| n % p != 0)).map(|n| n - size).collect::<Vec<_>>();
    let mut increments = vec![0; *buckets.last().unwrap() as usize];
    (0..buckets.len() - 1).for_each(|i| increments[buckets[i] as usize] = buckets[i + 1] - buckets[i]);
    GeneratorWheel { size, increments, sieve: vec![] }
}

impl Iterator for GeneratorWheel {
    type Item = isize;

    fn next(&mut self) -> Option<Self::Item> {
        let next = match self.sieve.last() {
            None => { 2 }
            Some(2) => {
                self.sieve.clear();
                3
            }
            last => {
                if *last.unwrap() < self.size {
                    (last.unwrap() + 2..).step_by(2).find(|&candidate| prime_sieve(candidate, &self.sieve)).unwrap()
                } else {
                    let (mut candidate, mut index) = (*last.unwrap(), (last.unwrap() % self.size));
                    loop {
                        let increment = self.increments[index as usize];
                        candidate += increment;
                        if prime_sieve(candidate, &self.sieve) {
                            break candidate;
                        }
                        index += increment;
                        if index > self.size { index = 1 }
                    }
                }
            }
        };

        self.sieve.push(next);
        Some(next)
    }
}

// --- //

/// closure that generates primes prime numbers, starting with the one below N.
pub struct PrimesLessThan {
    pub n: isize
}

pub fn primes_less_than(n: isize) -> PrimesLessThan {
    PrimesLessThan { n: if is_even(n) { n - 1 } else { n } }
}

impl Iterator for PrimesLessThan {
    type Item = isize;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            self.n -= 2;
            if miller_rabin(self.n) {
                return if self.n > 1 { Some(self.n) } else { None };
            }
        }
    }
}

// --- //

/// closure that generates primorals, the product of the smallest primes
pub struct Primoral {
    generator: GeneratorTrialDivision,
    product: isize
}

pub fn primoral() -> Primoral {
    Primoral { generator: generator_trial_division(), product: 1 }
}

impl Iterator for Primoral {
    type Item = isize;

    fn next(&mut self) -> Option<Self::Item> {
        self.product *= self.generator.next().unwrap();
        Some(self.product)
    }
}

// --- //

pub fn miller_rabin(value: isize) -> bool {
    value != 1 && if value < MR_THRESHOLD { MR_BASE_FAST } else { MR_BASE }.iter().all(|&b| value <= b || miller_rabin_pass(b, value))
}

fn miller_rabin_pass(b: isize, value: isize) -> bool {
    let s = (value - 1).trailing_zeros() as isize;
    let d = (value - 1) >> s;
    let mut a = power_modulo(b, d, value);

    if a == 1 {
        return true;
    }
    for _ in 0..s - 1 {
        if a == value - 1 {
            return true;
        }
        if a == 1 {
            return false;
        }
        a = power_modulo(a, 2, value);
    }
    a == value - 1
}
